kaiax/reward/config.go

```diff
@@ -0,0 +1,139 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package reward
+
+import (
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/consensus/istanbul"
+	"github.com/kaiachain/kaia/kaiax/gov"
+	"github.com/kaiachain/kaia/params"
+)
+
+var big100 = big.NewInt(100)
+
+type RewardConfig struct {
+	Rules      params.Rules
+	Rewardbase common.Address // Proposer's reward receipient address
+
+	IsSimple      bool              // istanbul.policy != WeightedRandom in which case simple rules are used
+	UnitPrice     *big.Int          // governance.unitprice
+	MintingAmount *big.Int          // reward.mintingamount
+	MinimumStake  *big.Int          // reward.minimumstake
+	DeferredTxFee bool              // reward.deferredtxfee
+	RewardRatio   *RewardRatio      // reward.ratio
+	Kip82Ratio    *RewardKip82Ratio // reward.kip82ratio
+}
+
+func NewRewardConfig(chainConfig *params.ChainConfig, govModule gov.GovModule, header *types.Header) (*RewardConfig, error) {
+	rc := &RewardConfig{}
+
+	rc.Rules = chainConfig.Rules(header.Number)
+	rc.Rewardbase = header.Rewardbase
+
+	paramset := govModule.EffectiveParamSet(header.Number.Uint64())
+	rc.IsSimple = paramset.ProposerPolicy != uint64(istanbul.WeightedRandom)
+	rc.UnitPrice = new(big.Int).SetUint64(paramset.UnitPrice)
+	rc.MintingAmount = new(big.Int).Set(paramset.MintingAmount)
+	rc.MinimumStake = new(big.Int).Set(paramset.MinimumStake)
+	rc.DeferredTxFee = paramset.DeferredTxFee
+
+	if ratio, err := NewRewardRatio(paramset.Ratio); err != nil {
+		return nil, err
+	} else {
+		rc.RewardRatio = ratio
+	}
+
+	if kip82Ratio, err := NewRewardKip82Ratio(paramset.Kip82Ratio); err != nil {
+		return nil, err
+	} else {
+		rc.Kip82Ratio = kip82Ratio
+	}
+
+	return rc, nil
+}
+
+// Parsed and validated reward.ratio parameter.
+type RewardRatio struct {
+	g int64 // Validators (GC)
+	x int64 // Fund1 (KIF, KFF, KGF, PoC)
+	y int64 // Fund2 (KEF, KCF, KIR)
+}
+
+func NewRewardRatio(ratio string) (*RewardRatio, error) {
+	parts := strings.Split(ratio, "/")
+	if len(parts) != 3 {
+		return nil, errMalformedRewardRatio(ratio)
+	}
+
+	g, err1 := strconv.ParseInt(parts[0], 10, 64)
+	x, err2 := strconv.ParseInt(parts[1], 10, 64)
+	y, err3 := strconv.ParseInt(parts[2], 10, 64)
+	if err1 != nil || err2 != nil || err3 != nil || g+x+y != 100 || g < 0 || x < 0 || y < 0 {
+		return nil, errMalformedRewardRatio(ratio)
+	}
+	return &RewardRatio{g: g, x: x, y: y}, nil
+}
+
+// Split splits the amount into three parts according to the ratio.
+func (r *RewardRatio) Split(amount *big.Int) (*big.Int, *big.Int, *big.Int) {
```

Comment:
Don't we need to care about precision issue? `gAmount + xAmount + yAmount` will be different from `amount` by maximum 2. 

But if we use remainder, the amount would be different under same ratio by maximum 1.

Reply:
Remainders will be handled outside Split(). e.g.
```go
	proposer, kif, kef := config.RewardRatio.Split(totalReward)
	ratioRemainder := calcRemainder(totalReward, proposer, kif, kef)
```

Reply:
Oh, I missed that. Sorry for the hussle.